src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlDateAddTranslator.cs

```diff
@@ -29,39 +66,52 @@ public class NpgsqlDateAddTranslator : IMethodCallTranslator
         };
 
         /// <summary>
-        ///     Translates the given method call expression.
+        /// The backend version to target.
         /// </summary>
-        /// <param name="methodCallExpression">The method call expression.</param>
-        /// <returns>
-        ///     A SQL expression representing the translated MethodCallExpression.
-        /// </returns>
+        [NotNull] readonly Version _postgresVersion;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="NpgsqlDateAddTranslator"/> class.
+        /// </summary>
+        /// <param name="postgresVersion">The backend version to target.</param>
+        public NpgsqlDateAddTranslator([NotNull] Version postgresVersion)
+            => _postgresVersion = Check.NotNull(postgresVersion, nameof(postgresVersion));
+
+        /// <inheritdoc />
         public virtual Expression Translate(MethodCallExpression methodCallExpression)
         {
-            if (!_methodInfoDatePartMapping.TryGetValue(methodCallExpression.Method, out var datePart))
+            // This translation is only supported for PostgreSQL 9.4 or higher.
+            if (_postgresVersion < MinimumSupportedVersion)
+                return null;
+
+            if (!MethodInfoDatePartMapping.TryGetValue(methodCallExpression.Method, out var datePart))
                 return null;
 
             // Ideally we would use Expression.Add() to have a simple plus-sign expression generated for us.
             // But to represent an interval we'd need to use a TimeSpan, which does not represent months/years.
             // We could use the provider-specific NpgsqlTimeSpan but it's not currently supported and is somewhat
             // ugly, so we just generate the expression ourselves with CustomBinaryExpression
 
-            var amountToAdd = methodCallExpression.Arguments.First();
+            if (!(methodCallExpression.Object is Expression instance) ||
```

Comment:
I like the new pattern matching syntax, but I'm not sure I like it as much as you :)

Unless I'm mistaken, if we're here, we already know that `methodCallExpression.Object` is non-null (thanks to the lookup above), that its type is `Expression`, and that there's one argument. So the conditions here are completely redundant, and seem to be used only to define new local variables, no?

If that's the case, I'd rather just define local variables. This makes the code easier to reason about, and doesn't make us doubt that maybe there's a chance of a null being there, etc...

I'm not sure I like this form of programming: the type of `methodCallExpression.Object` is `Expression`, and unless

Reply:
I am a bit enamoured with the pattern matching...but you're right that this one is a bit over the top.

However, I think we still want a (simple) null check on both the instance and the argument. While `DateTime` and `DateTimeOffset` are structs, it's always possible that an earlier translator/visitor may have have forwarded a null reference.